compiler/src/linking/link.re

```diff
@@ -0,0 +1,529 @@
+  | Select =>
+    globalize_names(local_names, Expression.select_get_if_true(expr));
+    globalize_names(local_names, Expression.select_get_if_false(expr));
+    globalize_names(local_names, Expression.select_get_condition(expr));
+  | Drop => globalize_names(local_names, Expression.drop_get_value(expr))
+  | Return => globalize_names(local_names, Expression.return_get_value(expr))
+  | MemorySize => ()
+  | MemoryGrow =>
+    globalize_names(local_names, Expression.memory_grow_get_delta(expr))
+  | Unreachable => ()
+  | Pop
+  | RefNull
+  | RefIs
+  | RefFunc
+  | RefEq
+  | Try
+  | Throw
+  | Rethrow => failwith("NYI")
+  | TupleMake =>
+    let num_operands = Expression.tuple_make_get_num_operands(expr);
+    for (i in 0 to num_operands - 1) {
+      globalize_names(
+        local_names,
+        Expression.tuple_make_get_operand_at(expr, i),
+      );
+    };
+  | TupleExtract =>
+    globalize_names(local_names, Expression.tuple_extract_get_tuple(expr))
+  | I31New
+  | I31Get
+  | CallRef
+  | RefTest
+  | RefCast
+  | BrOn
+  | RttCanon
+  | RttSub
+  | StructNew
+  | StructGet
+  | StructSet
+  | ArrayNew
+  | ArrayGet
+  | ArraySet
+  | ArrayLen
+  | RefAs => failwith("NYI")
+  };
+};
+
+let table_offset = ref(0);
+let module_id = ref(0);
+
+let link_all = (linked_mod, dependencies) => {
+  table_offset := 0;
+
+  let link_one = dep => {
+    let local_names: Hashtbl.t(string, string) = Hashtbl.create(128);
+
+    let wasm_mod = Hashtbl.find(modules, dep);
+
+    let num_globals = Global.get_num_globals(wasm_mod);
+    for (i in 0 to num_globals - 1) {
+      let global = Global.get_global_by_index(wasm_mod, i);
+      if (is_global_imported(global)) {
+        let imported_module = Import.global_import_get_module(global);
+        if (is_grain_module(imported_module)) {
+          let imported_name = Import.global_import_get_base(global);
+          let internal_name = Global.get_name(global);
+          let new_name =
+            Hashtbl.find(
+              Hashtbl.find(exported_names, imported_module),
+              imported_name,
+            );
+          Hashtbl.add(local_names, internal_name, new_name);
+        } else {
+          let imported_name = Import.global_import_get_base(global);
+          let internal_name = Global.get_name(global);
+          let new_name = gensym(internal_name);
+          Hashtbl.add(local_names, internal_name, new_name);
+
+          if (imported_module == "grainRuntime") {
+            let value =
+              switch (imported_name) {
+              | "relocBase" =>
+                Expression.const(
+                  wasm_mod,
+                  Literal.int32(Int32.of_int(table_offset^)),
+                )
+              | "moduleRuntimeId" =>
+                incr(module_id);
+                Expression.const(
+                  wasm_mod,
+                  Literal.int32(Int32.of_int(module_id^)),
+                );
+              | _ => assert(false)
```

Comment:
Why `assert(false)` instead of a `failwith` or `raise`?

Reply:
Draft PR 😛